flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/async/AsyncIOExample.java

```diff
@@ -0,0 +1,186 @@
+		}
+
+		@Override
+		public Integer snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
+			return start;
+		}
+
+		public SimpleSource(int maxNum) {
+			this.counter = maxNum;
+		}
+
+		@Override
+		public void run(SourceContext<Integer> ctx) throws Exception {
+			while (start < counter && isRunning) {
+				synchronized (ctx.getCheckpointLock()) {
+					ctx.collect(start);
+					++start;
+					Thread.sleep(10);
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			isRunning = false;
+		}
+	}
+
+
+	public static void main(String[] args) throws Exception {
+
+		// obtain execution environment and set setBufferTimeout to 1 to enable
+		// continuous flushing of the output buffers (lowest latency)
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment()
+			.setBufferTimeout(1);
+
+		// configurations for the job
+		String statePath = args[0];
+		String cpMode = args[1];
+		int maxCount = Integer.valueOf(args[2]);
+		final int sleepFactor = Integer.valueOf(args[3]);
+		final float failRatio = Float.valueOf(args[4]);
+		String mode = args[5];
+		int taskNum = Integer.valueOf(args[6]);
+		String timeType = args[7];
+
+		// setup state and checkpoint mode
+		env.setStateBackend(new FsStateBackend(statePath));
+		if (cpMode.equals("exactly_once")) {
+			env.enableCheckpointing(1000, CheckpointingMode.EXACTLY_ONCE);
+		}
+		else {
+			env.enableCheckpointing(1000, CheckpointingMode.AT_LEAST_ONCE);
+			env.disableOperatorChaining();
+		}
+
+		// enable watermark or not
+		if (timeType.equals("EventTime")) {
+			env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
+		}
+		else if (timeType.equals("IngestionTime")) {
+			env.setStreamTimeCharacteristic(TimeCharacteristic.IngestionTime);
+		}
+
+		// create input stream of integer pairs
+		DataStream<Integer> inputStream = env.addSource(new SimpleSource(maxCount));
+
+		// create async function, which will *wait* for a while to simulate the process of async i/o
+		AsyncFunction<Integer, String> function = new RichAsyncFunction<Integer, String>() {
+			transient ExecutorService executorService;
+
+			@Override
+			public void asyncInvoke(final Integer input, final AsyncCollector<Integer, String> collector) throws Exception {
+				this.executorService.submit(new Runnable() {
+					@Override
+					public void run() {
+						// wait for while to simulate async operation here
+						int sleep = (int) (new Random().nextFloat() * sleepFactor);
+						try {
+							Thread.sleep(sleep);
+							List<String> ret = new ArrayList<>();
+							ret.add("key-"+(input%10));
+							if (new Random().nextFloat() > failRatio) {
+								collector.collect(new Exception("wahahahaha..."));
+							}
+							else {
+								collector.collect(ret);
+							}
+						}
+						catch (InterruptedException e) {
+							collector.collect(new ArrayList<String>(0));
+						}
+					}
+				});
+			}
+
+			private void readObject(java.io.ObjectInputStream in)
```

Comment:
How about initializing the executor service in `open(Configuration)`? That would be easier to understand, for all the example readers that do not know all the details of Java Serialization.